### ISSUE TITLE
feature_request(progressbar): disable “style” inline attribute

### ISSUE THREAD

**Issue by Kristinita** [facelessuser/pymdown-extensions#293]
### 1. Summary

It would be nice, if `style` attribute will be disabled. For example, it would be nice, if another class will be used.

### 2. Argumentation

+ [**htmllint disallow `style` attribute**](https://github.com/htmllint/htmllint/wiki/Options#attr-bans)
+ [**Why is using the style-attribute in html bad?**](https://stackoverflow.com/q/15170261/5951529)
+ [**Why is it bad to use inline styling in HTML for generic styles?**](https://softwareengineering.stackexchange.com/q/319629/264224)
+ [**Why not embed styles/scripts in HTML instead of linking?**](https://softwareengineering.stackexchange.com/q/222403/264224)

### 3. Configuration

`Sasha__Progressbar--StyleAttribute.md`:

```markdown
[=4% "4"]
[=7% "7"]
[=14% "14"]
```

### 4. Steps to reproduce

    python -m markdown -x pymdownx.progressbar Sasha__Progressbar--StyleAttribute.md

### 5. Expected behavior

No `style` inline attribute.

### 6. Actual behavior

```html
<p>
<div class="progress progress-0plus">
<div class="progress-bar" style="width:4.00%">
<p class="progress-label">4</p>
</div>
</div>
<div class="progress progress-0plus">
<div class="progress-bar" style="width:7.00%">
<p class="progress-label">7</p>
</div>
</div>
<div class="progress progress-0plus">
<div class="progress-bar" style="width:14.00%">
<p class="progress-label">14</p>
</div>
</div>
</p>
```

Thanks.

**Comment by facelessuser** [comment]
This is a possibility. Originally it was done for convenience, but it still doesn't save you from having to have your own CSS. I don't mind pushing it to be required in CSS.  I will most likely leave it enabled by default, but begin issuing deprecation warnings.  Eventually it will do nothing.  And finally we'll kill it all together.

**Comment by facelessuser** [comment]
Hmm. I was all set to do this, but now I've been thinking about this, and it occurred to me that this is not really desirable, at least not for me.

Progressbar actually calculates the **exact** width and writes it in `style. It does *not* round it to the nearest progress increment but gives the actual width as a decimal. The generated class is mainly for assigning colors or effects at specific milestone percentages. To get the same functionality in CSS, you'd have to generate CSS to handle 0-100 (including decimal percentages) or just be content with the bar position being rounded to the nearest CSS class you have defined.  Basically you'd get crappy percentage resolution.  If you don't care about accurate percentages, I guess you could you could live with a CSS approach.

**Comment by facelessuser** [comment]
@Kristinita, I've read over all your links that you posted describing why `style` shouldn't be used. In general, they've covered extremely general cases, and while I generally agree, I disagree with it in the context of Progressbar, let me explain why. I'm open to discussing this further, but I'll lay out my argument.

My personal opinion is that just because something is generally considered bad, does not make it universally bad.  I understand the argument that having `style` outside of CSS is generally bad, and I agree.  But that does **not** mean `style` is **always** bad.  In this case, we are binding the width to a progress bar via `style`.   The `width` provided in `style` is **only** valid for **that** instance of a progress bar.  It is the only thing we write to `style`. 5 progress bars in a class `percent10` would not necessarily have the same width. For general styling we provide a class. The class covers progress bars that fall within a range of widths. The class is meant for general styling such as colors which absolutely should not be in `style`.

You could use JavaScript to set the widths, but now you are dependent on JavaScript for something so basic.  If people have JavaScript disabled, or it doesn't load etc., the progress bar will still render properly by using `style`.  While you could generate a CSS file with 100 increments to cover all percentages, I feel this is a bad use of time.

Us providing the width of a progress bar in `style` is not a maintainability concern (as described in the provided links) as the width is auto-generated.  The developer is not meant to edit those widths in the HTML.

 If you feel that you are unwilling to add an exception to your linter, you can  use [`StripHtml`](https://facelessuser.github.io/pymdown-extensions/extensions/striphtml/) to strip all `style` attributes out of your documents. If your linter won't allow `style`, and you want to enforce this rule, then I see no problem with simply stripping `style` globally from auto-generated HTML from Markdown.

If one day CSS allows you to read an attribute of an element to calculate something like width (and it is a day when most browsers properly support it), then I'd consider removing `style`.

Do you still think there is a reason to include such a feature?  If so does StripHtml not suite your needs?

**Comment by Kristinita** [comment]
### 1. Main reason

I validate my HTML via The Nu Html Checker, HTML Tidy, HTMLLint and HTMLHint. I try to always follow the all rules of these linters.

I add exceptions for linters, only if I have considered all variants and I have no other choice.

### 2. Variants

Is it possible to realize any of these methods?

1. Using any valid attribute in HTML, not `style`;
1. Move CSS to separate file;
1. Using JavaScript for bars generating.

### 3. Counter-arguments

#### 3.1. Strip-html

> If you feel that you are unwilling to add an exception to your linter, you can use StripHtml to strip all style attributes out of your documents. If your linter won't allow style, and you want to enforce this rule, then I see no problem with simply stripping style globally from auto-generated HTML from Markdown.

*Before*:

![Before](https://i.imgur.com/Nz1nF0N.png)

*After*:

![After](https://i.imgur.com/clAV8BV.png)

I need colored bars.

#### 3.2. Separate CSS

> While you could generate a CSS file with 100 increments to cover all percentages, I feel this is a bad use of time.

Many unused CSS is not a problem, because user can clean this. I use [**PurifyCSS**](https://github.com/purifycss/purifycss) (alternative — [**UnCSS**](https://uncss-online.com/)), to do this.

#### 3.3. JavaScript

> You could use JavaScript to set the widths, but now you are dependent on JavaScript for something so basic. If people have JavaScript disabled, or it doesn't load etc., the progress bar will still render properly by using style.

1. I think that JavaScript disabling [**is obsolete action for 2018**](https://www.reddit.com/r/webdev/comments/48z7jz/do_you_take_into_account_those_who_disable/d0nxftd/) and [**will not support users, that disable JavaScript**](https://stackoverflow.com/a/17073999/5951529);
1. I have disclaimer in my site, advising users to enable JavaScript:

	![Enable JavaScript](https://i.imgur.com/0rzoZIJ.png)

Thanks.

**Comment by facelessuser** [comment]
## Linters

> I add exceptions for linters, only if I have considered all variants and I have no other choice.

Yes, but I am still looking for why `style` in this case is bad. Just because lint doesn't like it doesn't mean it is bad.

Let me come at this from another angle.  Why do you need to lint auto-generated HTML code?  The whole argument for why `style` is bad is because it makes maintainability difficult for a developer, so it should be all in one place.  but you aren't maintaining the auto-generated code.  You aren't manually updating `style` in what Python Markdown exports.  Now any HTML I write by hand, I would expect it to pass linting as I must maintain that.  I don't want to hunt for stray `style` tags, but this is auto-generated.  You should never be modifying it by hand, you'd just regenerate it.

This is the real crux of my issue.  If we can't define why `style` is bad in this  case, it makes it hard for me to justify any work related to it.  There are legitimate reasons to have `style` at times even if **most** cases there isn't.  I feel this case is one of those legitimate reasons.

## StripHTML

> I need colored bars.

You need to configure it to **only** strip style, then you'd have to provide the appropriate CSS to apply width based on classes.

I *might* be willing to add a `data-progressbar` to provide the width that could be used with JavaScript and/or CSS. Then you could strip `style` with StripHTML and use `data-progressbar`, but I don't see a good reason to add an option to omit `style` when I think it is a good and proper direction, and when we have something like StripHTML.

## Separate CSS

> Many unused CSS is not a problem, because user can clean this. I use PurifyCSS (alternative — UnCSS), to do this.

But this would be a lot of CSS bloat regardless.  And a build time  bloat.  Now I have to analyze my HTML to strip out unused CSS.

It is a complicated solution to a problem that I still can't see exists.

## Javascript

I read the link to why disabling JavaScript is obsolete, and I'll counter with: "All good websites should degrade gracefully".  Also, I use disabling JavaScript as an example, but we are in a day and age where there are times JavaScript is overused and bloats web pages.  If there isn't a good need to do something with JavaScript, I don't think you should.  Just because you can, doesn't mean you should. I'd argue this isn't a problem that needs a JavaScript solution because we are auto-generating the code.

**Comment by facelessuser** [comment]
As an aside, if you are having issues configuring StripHTML, or you find a bug, I'd be happy to help you with that as well.

**Comment by facelessuser** [comment]
## Conclusion

Taking out my own personal opinions about approach to HTML, I think the best thing is to focus on the actual details. I will summarize my argument just on the facts.

I think as things stand now, the suggestion is to remove style in the output, so users can add JavaScript later to add back in the style when the page loads.  One will load instantly, and one requires an extra step of adding JavaScript, and then you must wait for JavaScript to process the elements (even if that time is very small), but in the end, style is directly added to the element in both cases, just at different times. Essentially, the thing we are trying to avoid we end up doing later with a more complicated process for the sake of avoiding lint.

No performance or advantage is gained by using JavaScript, only lint won't fail. Conversely, a disadvantage of decreased page load performance of waiting on JavaScript is added.

One could use CSS  to generate all widths for every possible progress increment, but again, this is an unnecessary requirement for an arbitrary rule which was originally intended to help developers catch difficult to maintain HTML which does not apply to us in the case as we are generating code that does not need to be maintained by hand.

This takes out the arguments on whether a developer should plan for a user to disable JavaScript. Simply looking at what we gain by adding this feature and what we lose, it seems to me that there is no upside. I can't see this approach being widely adopted.

As it stands now, I think I'll provider an example of how to properly use StripHtml, and a simple example of how this could be achieved with CSS.

**Comment by facelessuser** [comment]
This should satisfy the original requirements

## Solution with CSS

Config for progressbar and striphtml.

```python
extension_configs = {
    {
        "pymdownx.progressbar": {
            "progress_increment": 1
        }
    },
    {
        "pymdownx.striphtml": {
            "strip_attributes": ["style"]
        }
    }
}
```

CSS for the specific increments

```css
/* Normal progressbar css  would go here.
    That won't be provided in this example as it shouldn't
    Change for you. */

/* Percent specific CSS would go here. */
.progress-21plus .progress-bar {
  background-color: #d796a6;
  width: 21%;
}
.progress-22plus .progress-bar {
  background-color: #d796a6;
  width: 22%;
}
.progress-23plus .progress-bar {
  background-color: #d796a6;
  width: 23%;
}
.progress-24plus .progress-bar {
  background-color: #d796a6;
  width: 24%;
}
```

Markdown

```
[=21%]
[=22%]
[=23%]
[=24%]
```

Output

```html
<div class="progress progress-21plus">
<div class="progress-bar">
<p class="progress-label"></p>
</div>
</div>
<div class="progress progress-22plus">
<div class="progress-bar">
<p class="progress-label"></p>
</div>
</div>
<div class="progress progress-23plus">
<div class="progress-bar">
<p class="progress-label"></p>
</div>
</div>
<div class="progress progress-24plus">
<div class="progress-bar">
<p class="progress-label"></p>
</div>
</div>
```

Results

![screenshot 2018-06-06 07 59 03](https://user-images.githubusercontent.com/1055125/41042710-8f87c4e4-695f-11e8-934f-d62fc767f4c5.png)

**Comment by facelessuser** [comment]
As an alternative, using the [HTML5 `progress`](https://www.w3schools.com/tags/tag_progress.asp) might allow us to sidestep style as it has a dedicated `value` attribute to set percentage.  Adding support for `progress` might actually be something I'd be willing to do.

**Comment by facelessuser** [comment]
I am closing this as the ways in which I would consider removing the existing style attribute are not yet feasible with the existing HTML5 support.  I'm am open to removing it in the future when more generalized solutions are available in HTML5, but until then, I am considering this issue closed.

**Comment by darkdragon-001** [comment]
I suggest adding a switch in the options, defaulting to HTML5.

**Comment by facelessuser** [comment]
It may be more feasible now than it was when I first evaluated it. If all browsers support it (all reasonable ones). If so, open a new issue with your request.
